docs/design/app/workflows.md

```diff
@@ -47,4 +48,33 @@ An `Event` at a time is sent to the `prehandle` with a reference to the latest i
 4. **Verify Signatures.** The information prepared in the previous step is sent to the platform to validate the signatures.
 5. **Transaction Metadata.** The `TransactionMetadata` generated by the `PreTransactionHandler` is attached to the `SwirldsTransaction`.
 
-If all checks have been successful, the status of the created `TransactionMetadata` will be `OK`. Otherwise, the status is set to the response code providing the failure reason. If the workflow terminates early (either because the parsing step (1.) fails or an unexpected `Exception` occurs) an `ErrorTransactionMetadata` is attached to the `SwirldsTransaction` that contains the causing `Exception`.
+If all checks have been successful, the status of the created `TransactionMetadata` will be `OK`. 
+Otherwise, the status is set to the response code providing the failure reason. 
+If the workflow terminates early (either because the parsing step (1.) fails or an unexpected `Exception` occurs) an `ErrorTransactionMetadata` is attached to the `SwirldsTransaction` that contains the causing `Exception`.
+
+### Query Workflow
+
+The query workflow is by far the most complex workflow.
+Unlike transaction processing, it is not split into several phases, but covers the whole query from receiving the request until sending the response.
+Also, a query typically contains a `CryptoTransfer` to cover the costs.
+
+![Diagram of query worflow](images/Query%20Workflow.svg)
+
+When a query arrives at the HAPI-endpoint,  the byte-buffer that contains the query is sent to the query workflow.
+The gRPC-server is responsible for `Thread`-management.
+The query-workflow is single-threaded, but multiple calls can run in parallel.
+
+The query workflow consists of the following steps:
+
+1. **Parse and check header.** The query arrives as a byte-array. The required parts are parsed and the type of the query (header) is extracted and validated.
+2. **Check query throttles.** Throttling must be observed and checked as early as possible.
+3. If a payment is required:
+   1. **Validate the CryptoTransfer.** The contained `CryptoTransfer` is validated.
+   2. **Check account balances.** The accounts of all payers are checked.
+4. **Check permissions.** It is checked, if the requester is actually allowed to do the query.
```

Comment:
May be it is already discussed. Is there a reason to not do the `Permissions` and `Validity` after throttles in the picture ?

Reply:
This is more or less the order in which these things are done today. We did not discuss any changes (yet).

My guess is, that we want to ensure the payer can pay for the query before doing any further work.

Reply:
After our discussion last week, I updated the documentation accordingly.